### ISSUE TITLE
Smoke tests: avoid `retries` unless there is a good reason for

### ISSUE THREAD

**Issue by bpasero** [microsoft/vscode#155816]
https://github.com/microsoft/vscode/blob/62469028a6dce9db5b7c4639be7f447f4ab0e35c/test/smoke/src/areas/terminal/terminal.test.ts#L19-L20

https://github.com/microsoft/vscode/blob/62469028a6dce9db5b7c4639be7f447f4ab0e35c/test/smoke/src/areas/task/task.test.ts#L13-L14

If we have understood the reason for a retry, then we should document it, such as these 2 cases:

https://github.com/microsoft/vscode/blob/62469028a6dce9db5b7c4639be7f447f4ab0e35c/test/smoke/src/areas/terminal/terminal-input.test.ts#L28-L32

https://github.com/microsoft/vscode/blob/62469028a6dce9db5b7c4639be7f447f4ab0e35c/test/smoke/src/areas/workbench/data-loss.test.ts#L148-L155

In other words, we should not just retry without knowing why.

**Comment by Tyriar** [comment]
I agree, however I spent quite a while trying to get rid of flakiness in terminal tests. So the alternative is to continue to invest x hours and cause headaches with flaky builds happening again, or flip this switch and have great coverage with the rare retry. 

What we really need to live by this rule imo is to be able to have our CI tell us about flakiness _in retries_, so we can create smoke tests with retries enabled and deflake them over time. I looked into this before but I'm not very knowledgeable around interfacing with pipelines and I think we would need to add some hooks to mocha too. @lszomoru do you think this is something we could accomplish now as it's becoming more important?

**Comment by bpasero** [comment]
From my experience of having spent weeks to figure out flaky tests, I actually found surprising things that helped me understand sources of flake and made me built an extensive set of diagnostic tools for smoke tests (tracing, logging), even rewriting parts of the smoke test framework. So I think there is value.

Besides, for smoke tests a flaky condition means a user likely hits the same issue.

**Comment by Tyriar** [comment]
> Besides, for smoke tests a flaky condition means a user likely hits the same issue.

👍 problems with the Playwright runner are very rare in my experience so this is definitely true.

**Comment by jrieken** [comment]
> What we really need to live by this rule imo is to be able to have our CI tell us about flakiness in retries, so we can create smoke tests with retries enabled and deflake them over time.

My learning is that flakiness isn't a property of the environment (CI) but a property of the code that's being tested. Running a flaky test locally repeatedly (inside a loop to 1000 or whatever) will almost always show the failure locally.

**Comment by lszomoru** [comment]
@Tyriar, Azure DevOps has the concept of [flacky tests](https://docs.microsoft.com/en-us/azure/devops/pipelines/test/flaky-test-management?view=azure-devops#flaky-test-detection). I have enabled this for both the "VS Code" and "Code OSS" pipelines. What's interesting is that one can even [plug-in custom code](https://docs.microsoft.com/en-us/azure/devops/pipelines/test/flaky-test-management?view=azure-devops#flaky-test-detection) to detect whether a test is flacky or not. I have not done this before but it is something that we can explore.

**Comment by lszomoru** [comment]
@Tyriar, any concerns with closing this issue?

**Comment by Tyriar** [comment]
@lszomoru ideally the flakiness would be resolved but I've invested way too much time in this so I'm fine leaving it as is and closing this off. The terminal smoke tests are notoriously flaky as they have so many dependencies with the process launches. We have the build champ now to point out flaky tests that become a problem.
